pkg/filter/interpreters.go

```diff
@@ -0,0 +1,209 @@
+// scanner to read the appropriate data from a Reader.
+type JSONInterpreter struct{}
+
+var _ Interpreter = JSONInterpreter{}
+
+// Interpret implements Interpreter for JSONInterpreter
+func (JSONInterpreter) Interpret(data []byte,
+	fields map[string]interface{}) ([]byte, map[string]interface{}) {
+	var parsed map[string]interface{}
+	err := json.Unmarshal(data, &parsed)
+	if err != nil {
+		// if it wasn't json, just do nothing
+		return data, fields
+	}
+	for k, v := range parsed {
+		fields[k] = v
+	}
+	return nil, fields
+}
+
+// LastChanceInterpreter should be the last Interpreter in the chain. It
+// takes any remaining data bytes and escapes them into a string, and sets
+// the field _other to the result. We're assuming since this was intended
+// to be a log entry that the data is mostly string-like, but there is
+// the option to run an Escaper over it so that we don't try to print gibberish.
+// Interpret implements Interpreter for LastChanceInterpreter
+type LastChanceInterpreter struct {
+	Escaper func([]byte) string
+}
+
+var _ Interpreter = LastChanceInterpreter{}
+
+// Interpret implements Interpreter for LastChanceInterpreter
+func (i LastChanceInterpreter) Interpret(data []byte,
+	fields map[string]interface{}) ([]byte, map[string]interface{}) {
+	if len(data) != 0 {
+		if i.Escaper != nil {
+			fields["_other"] = i.Escaper(data)
+		} else {
+			fields["_other"] = string(data)
+		}
+	}
+	return nil, fields
+}
+
+// RequiredFieldsInterpreter simply copies its default fields into the
+// destination and then passes on its input data unexamined.
+// You can control whether these fields override existing fields
+// or are overridden by where this sits in the stack of interpreters.
+type RequiredFieldsInterpreter struct {
+	Defaults map[string]interface{}
+}
+
+var _ Interpreter = (*RequiredFieldsInterpreter)(nil)
+
+// Interpret implements Interpreter for RequiredFieldsInterpreter
+func (i *RequiredFieldsInterpreter) Interpret(data []byte,
+	fields map[string]interface{}) ([]byte, map[string]interface{}) {
+	for k, v := range i.Defaults {
+		fields[k] = v
+	}
+	return data, fields
+}
+
+// TendermintInterpreter looks at the specific keys specified
+// and attempts to interpret them further by parsing them for
+// things that look like "name: value". You'd generally want to
+// put this in the list after a JSONInterpreter has split the
+// file up.
+type TendermintInterpreter struct {
+	Keys []string
+}
+
+// NewTendermintInterpreter constructs a TendermintInterpreter with the
+// one field that is currently worth searching: _msg
+func NewTendermintInterpreter() Interpreter {
+	return &TendermintInterpreter{Keys: []string{"_msg"}}
+}
+
+var _ Interpreter = (*TendermintInterpreter)(nil)
+
+func findFields(v string, fields map[string]interface{}) map[string]interface{} {
+	// pattern for matching lines that have key: value
+	lpat := regexp.MustCompile(`^([A-Z][A-Za-z0-9]+):[ \t]*(.*[^{])$`)
```

Comment:
This regex won't match the literal "key: value", because it expects that the first character of a key is uppercase. It might also be good to allow underscores within the key. 

I'm not sure how to interpret the intent behind the close of this RE: `(.*[^{])$`: it matches anything _except_ if the line ends in an open curly brace, in which case, the whole pattern is rejected? I think that's what it does, but I'm a bit confused as to why it does that.

Reply:
I'll add more comments -- this regex was very specifically designed for the odd mechanism used by tendermint within logging for their blocks.